gnd/src/main/java/com/google/android/gnd/ui/editrecord/EditRecordViewModel.java

```diff
@@ -78,6 +86,47 @@
     this.showUnsavedChangesDialogEvents = new SingleLiveEvent<>();
     this.showErrorDialogEvents = new SingleLiveEvent<>();
     this.authManager = authenticationManager;
+    this.editRecordRequests = PublishSubject.create();
+    this.recordSaveRequests = PublishSubject.create();
+
+    disposeOnClear(
+            recordSaveRequests.zipWith(authManager.getUser(), this::saveRecord)
+            .switchMap(Result::wrapObservable)
+            .subscribe(Result.unwrap(record::setValue, this::onSaveRecordError)));
+
+    disposeOnClear(
+        editRecordRequests
+            .switchMapSingle(Result.mapSingle(this::createOrUpdateRecord))
+            .subscribe(Result.unwrap(this::onRecordSnapshot, this::onEditRecordError)));
+  }
+
+  private Single<Resource<Record>> createOrUpdateRecord(EditRecordRequest request) {
+    return request.getIsNew() ? createRecord(request) : updateRecord(request);
+  }
+
+  private Single<Resource<Record>> createRecord(EditRecordRequest request) {
+    return this.dataRepository
+        .createRecord(request.getProjectId(), request.getFeatureId(), request.getFormId())
+        .map(Resource::loaded)
+        .doFinally(this::onNewRecordLoaded);
```

Comment:
`doFinally` is called both on error and on success. Did you mean `doOnSuccess()`?

Reply:
Shoot, you're right, error on my part. I was trying to make the signatures more "correct". Really, I was just being a bit pedantic.

`onNewRecordLoaded` doesn't do anything to the record, so it's not necessary to pass it one. `doOnSuccess` passes the success value to the method provided as an argument, while `doFinally` does not--`doOnSuccess` suggests the method is dependent on the success value, the `doFinally` that it isn't. This is just a semantic gap in RX since there's no `doOnSuccess(void)` so there's no notion of "only do this action if this succeeds, but do it independently of the value"--actually, this probably is expressible in RX terms but entails superfluous transformations.

I was just trying to avoid it because a function that takes an argument and does nothing with it is sometimes suspicious to me.

Reply:
Some investigation suggests `Single` is an outlier of sorts as it's apparently only implemented by RX Java and its derivatives. 

`doOnSuccess` has to satisfy both the role of `doOnNext` and `doOnCompleted`--that's why it cares about the value :/

In any case, we can use an alternative here.